File: medium/1129/1129. Shortest Path with Alternating Colors.py
import collections
from enum import Enum
from typing import List


class Color(Enum):
  kInit = 0
  kRed = 1
  kBlue = 2


class Solution:
  def shortestAlternatingPaths(self, n: int, redEdges: List[List[int]], blueEdges: List[List[int]]) -> List[int]:
    res = [-1] * n
    graph = [[] for _ in range(n)]
    for u, v in redEdges:
      graph[u].append((v, Color.kRed))
    for u, v in blueEdges:
      graph[u].append((v, Color.kBlue))
    print(graph)
    queue = collections.deque([(0, Color.kInit)])
    step = 0
    while queue:
      for _ in range(len(queue)):
        node, previousColor = queue.popleft()
        # visited node
        if res[node] == -1:
          res[node] = step
        # traverse all the neighbors of node
        for i, (v, color) in enumerate(graph[node]):
          if v == -1 or previousColor == color:
            continue
          queue.append((v, color))
          # mark the node as visited
          graph[node][i] = (-1, color)
      step += 1
    return res


p = Solution()
p.shortestAlternatingPaths(3, [[0, 1]], [[1, 2]])

File: medium/1129/1129. Shortest Path with Alternating Colors.rs
impl Solution {
  pub fn shortest_alternating_paths(n: i32, red_edges: Vec<Vec<i32>>, blue_edges: Vec<Vec<i32>>) -> Vec<i32> {
    use std::collections::VecDeque;
    // enum Color {
    //   Red,
    //   Blue,
    //   Init,
    // }
    let mut res = vec![-1; n];
    let n = n as usize;
    let mut graph = vec![vec![]; n];

    for edge in red_edges {
      graph[edge[0] as usize].push((edge[1], 1));
    }
    for edge in blue_edges {
      graph[edge[0] as usize].push((edge[1], 2));
    }

    let mut queue: VecDeque<(i32, i32)> = VecDeque::new();

    queue.push_back((0, 0));


    let mut step = 0;
    while !queue.is_empty() {
      for _ in 0..queue.len() {
        let (node, color) = queue.pop_front().unwrap();

        if res[node as usize] == -1 {
          res[node as usize] = step;
        }

        let length: usize = graph[node as usize].len();

        for i in 0..length {
          let mut edge = graph[node as usize][i];
          // the color of the edge that node to num [num,color]
          if edge.0 == -1 || edge.1 == color {
            continue;
          }
          queue.push_back((edge.0, edge.1));
          graph[node as usize][i].0 = -1;
        }
      }
      step += 1;
    }
    return res;
  }
}